webapp/login.go

```diff
@@ -0,0 +1,255 @@
+)
+
+// Copyright 2019 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+func init() {
+	gob.Register(map[string]interface{}{})
+	// Need RegisterName - for local packages, Register appends main[0-9]{5}.User
+	gob.RegisterName("User", User{})
+}
+
+// User represents an authenticated GitHub user.
+type User struct {
+	GitHubHandle string
+	GithuhEmail  string
+}
+
+func loginHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := shared.NewAppEngineContext(r)
+	aeAPI := shared.NewAppEngineAPI(ctx)
+	if !aeAPI.IsFeatureEnabled("githubLogin") {
+		http.Error(w, "Feature not implemented", http.StatusNotImplemented)
+		return
+	}
+
+	user, token := getUserFromCookie(r)
+	returnURL := r.FormValue("return")
+	redirect := returnURL
+	log := shared.GetLogger(ctx)
+	if user == nil || token == nil {
+		log.Infof("Initiating a new user login.")
+		conf := getGithubOAuthConfig(ctx)
+		conf.RedirectURL = getCallbackURI(returnURL, r)
+		state, err := generateRandomState(32)
+		if err != nil {
+			log.Errorf("Failed to generate random state")
+			http.Error(w, "Error creating a random state for login", http.StatusInternalServerError)
+			return
+		}
+		redirect = conf.AuthCodeURL(state, oauth2.AccessTypeOnline)
+		setState(ctx, state, w)
+		log.Infof("OAuthing with github and returning to %s", returnURL)
+	} else {
+		if redirect == "" {
+			redirect = "/"
+		}
+		log.Infof("User %s is logged in", user.GitHubHandle)
+	}
+
+	http.Redirect(w, r, redirect, http.StatusTemporaryRedirect)
+}
+
+func oauthHandler(w http.ResponseWriter, r *http.Request) {
+	oauthToken := r.FormValue("code")
+	if oauthToken == "" {
+		http.Error(w, "No token or username provided", http.StatusBadRequest)
+		return
+	}
+
+	ctx := shared.NewAppEngineContext(r)
+	conf := getGithubOAuthConfig(ctx)
+	token, err := conf.Exchange(ctx, oauthToken)
+	log := shared.GetLogger(ctx)
+	if err != nil {
+		log.Errorf("Invalid OAuth2 token: %s", err.Error())
+		http.Error(w, "Invalid OAuth2 token", http.StatusBadRequest)
+		return
+	}
+
+	oauthClient := oauth2.NewClient(ctx, oauth2.StaticTokenSource(token))
+	client := github.NewClient(oauthClient)
+	if err != nil {
+		log.Errorf("Failed to get GitHub client")
+		http.Error(w, "Error fetching user", http.StatusInternalServerError)
+		return
+	}
+	ghUser, _, err := client.Users.Get(ctx, "") // Empty string => Authenticated user.
+	if err != nil || ghUser == nil {
+		log.Errorf("Failed to get authenticated user")
+		http.Error(w, "Failed to get authenticated user", http.StatusBadRequest)
+		return
+	}
+	user := &User{
+		GitHubHandle: ghUser.GetLogin(),
+		GithuhEmail:  ghUser.GetEmail(),
+	}
+	setSession(ctx, user, &token.AccessToken, w)
+	log.Infof("User %s logged in", user.GitHubHandle)
+
+	ret := r.FormValue("return")
+	encodedState := r.FormValue("state")
+	stateFromCookie := getState(r)
+	if encodedState != stateFromCookie {
```

Comment:
We should verify the state as the very first step.

In the current implementation, if the states don't match, we'd already set session cookies via `setSession`, and `http.Error` does not unset those cookies, so the user would see a 500 error page but with their cookies set -- in other words, if they refresh the homepage, they'd be logged in.

Reply:
Good catch!